from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QApplication, QMainWindow, QMessageBox, QAbstractItemView
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QFont
from view.FormCadastrarProduto import Ui_FormCadastrarProduto
from view.FormAlterarProduto import Ui_FormAlterarProduto
from controller.ProdutoCTR import ProdutoCTR

class Ui_FormProdutos(object):
    def click_btn_cadastrar_produto(self):
        self.formcadastrarproduto = QMainWindow()
        self.ui = Ui_FormCadastrarProduto()
        self.ui.setupUi(self.formcadastrarproduto)
        self.formcadastrarproduto.show()

    def click_btn_alterar_produto(self):
        try:
            linha = self.tableProduto.currentItem().row()
            codprod = self.tableProduto.item(linha, 0).text()
            descricao = self.tableProduto.item(linha, 1).text()
            tipo = self.tableProduto.item(linha, 2).text()
            preco = float(self.tableProduto.item(linha, 3).text())
            self.formalterarproduto = QMainWindow()
            self.ui = Ui_FormAlterarProduto()
            self.ui.setupUi(self.formalterarproduto)
            self.formalterarproduto.show()
            self.ui.preencher_campos(codprod, descricao, tipo, preco)
        except:
            msg = ("Selecione um produto para alterar.")
            dlg = QMessageBox(None)
            dlg.setWindowTitle("Erro")
            dlg.setIcon(QMessageBox.Critical)
            font = QFont()
            font.setFamily("Arial")
            font.setPointSize(15)
            dlg.setFont(font)
            dlg.setWindowIcon(QtGui.QIcon("error.png"))
            dlg.setText(msg)
            dlg.exec_()



    def click_btn_excluir_produto(self):
        try:
            linha = self.tableProduto.currentItem().row()
            Codprod = (int(self.tableProduto.item(linha, 0).text()))
            descricao = self.tableProduto.item(linha, 1).text()
            msg = QMessageBox(None)
            msg.setWindowTitle("Confirmação")
            msg.setWindowIcon(QtGui.QIcon("question.png"))
            msg.setIcon(QMessageBox.Question)
            text = ("Deseja excluir o produto: \n"
                        "{} ?".format(descricao))
            msg.setText(text)
            msg.setStandardButtons(QMessageBox.Ok | QMessageBox.Cancel)
            op = msg.exec_()
            if (op == QMessageBox.Ok):
                ProdutoCTR.Excluir_Produto(Codprod)
            elif (op == QMessageBox.Cancel):
                msg = QMessageBox(None)
                msg.setWindowTitle("Cancelado")
                msg.setWindowIcon(QtGui.QIcon("error.png"))
                msg.setIcon(QMessageBox.Critical)
                msg.setText("Exclusão cancelada.")
                msg.exec_()
            self.click_btn_pesquisar_produto()


        except:
            msg = ("Selecione um produto para excluir.")
            dlg = QMessageBox(None)
            dlg.setWindowTitle("Erro")
            dlg.setIcon(QMessageBox.Critical)
            font = QFont()
            font.setFamily("Arial")
            font.setPointSize(15)
            dlg.setFont(font)
            dlg.setWindowIcon(QtGui.QIcon("error.png"))
            dlg.setText(msg)
            dlg.exec_()

    def click_btn_pesquisar_produto(self):
        aux = self.editPesquisaProduto.text()
        if (aux == ''):
            result = ProdutoCTR.Pesquisar_Todos_Produtos()
            self.tableProduto.setRowCount(0)
            for num_linha, linha_conteudo in enumerate(result):
                self.tableProduto.insertRow(num_linha)
                for num_coluna, coluna_conteudo in enumerate(linha_conteudo):
                    self.tableProduto.setItem(num_linha, num_coluna, QtWidgets.QTableWidgetItem(str(coluna_conteudo)))
        elif (aux!= '' ):
            nome = self.editPesquisaProduto.text()
            result = ProdutoCTR.Pesquisar_Produto(nome)
            self.tableProduto.setRowCount(0)
            for num_linha, linha_conteudo in enumerate(result):
                self.tableProduto.insertRow(num_linha)
                for num_coluna, coluna_conteudo in enumerate(linha_conteudo):
                    self.tableProduto.setItem(num_linha, num_coluna, QtWidgets.QTableWidgetItem(str(coluna_conteudo)))

    def setupUi(self, FormProdutos):
        FormProdutos.setObjectName("FormProdutos")
        FormProdutos.resize(661, 471)
        icon = QtGui.QIcon()
        icon.addPixmap(QtGui.QPixmap("listar produtos.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        FormProdutos.setWindowIcon(icon)
        self.centralwidget = QtWidgets.QWidget(FormProdutos)
        self.centralwidget.setObjectName("centralwidget")
        self.label = QtWidgets.QLabel(self.centralwidget)
        self.label.setGeometry(QtCore.QRect(30, 20, 51, 21))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.label.setFont(font)
        self.label.setObjectName("label")
        self.editPesquisaProduto = QtWidgets.QLineEdit(self.centralwidget)
        self.editPesquisaProduto.setGeometry(QtCore.QRect(90, 20, 371, 21))
        self.editPesquisaProduto.setText("")
        self.editPesquisaProduto.setObjectName("editPesquisaProduto")
        self.tableProduto = QtWidgets.QTableWidget(self.centralwidget)
        self.tableProduto.setGeometry(QtCore.QRect(10, 60, 641, 301))
        self.tableProduto.setSelectionMode(QAbstractItemView.SingleSelection)
        self.tableProduto.setTextElideMode(Qt.ElideRight)
        self.tableProduto.setWordWrap(False)
        self.tableProduto.setSortingEnabled(False)
        self.tableProduto.horizontalHeader().setDefaultAlignment(Qt.AlignHCenter | Qt.AlignVCenter |
                                                          Qt.AlignCenter)
        self.tableProduto.setSelectionBehavior(QAbstractItemView.SelectRows)
        self.tableProduto.horizontalHeader().setStretchLastSection(True)
        self.tableProduto.setEditTriggers(QtWidgets.QAbstractItemView.NoEditTriggers)
        self.tableProduto.setRowCount(0)
        self.tableProduto.setObjectName("tableProduto")
        self.tableProduto.setColumnCount(4)
        self.tableProduto.setColumnWidth(1, 200)
        item = QtWidgets.QTableWidgetItem()
        item.setTextAlignment(QtCore.Qt.AlignCenter)
        font = QtGui.QFont()
        font.setBold(True)
        font.setWeight(75)
        item.setFont(font)
        self.tableProduto.setHorizontalHeaderItem(0, item)
        item = QtWidgets.QTableWidgetItem()
        item.setTextAlignment(QtCore.Qt.AlignCenter)
        font = QtGui.QFont()
        font.setBold(True)
        font.setWeight(75)
        item.setFont(font)
        self.tableProduto.setHorizontalHeaderItem(1, item)
        item = QtWidgets.QTableWidgetItem()
        item.setTextAlignment(QtCore.Qt.AlignCenter)
        font = QtGui.QFont()
        font.setBold(True)
        font.setWeight(75)
        item.setFont(font)
        self.tableProduto.setHorizontalHeaderItem(2, item)
        item = QtWidgets.QTableWidgetItem()
        item.setTextAlignment(QtCore.Qt.AlignCenter)
        font = QtGui.QFont()
        font.setBold(True)
        font.setWeight(75)
        item.setFont(font)
        self.tableProduto.setHorizontalHeaderItem(3, item)
        self.layoutWidget = QtWidgets.QWidget(self.centralwidget)
        self.layoutWidget.setGeometry(QtCore.QRect(470, 15, 166, 31))
        self.layoutWidget.setObjectName("layoutWidget")
        self.horizontalLayout = QtWidgets.QHBoxLayout(self.layoutWidget)
        self.horizontalLayout.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.btnPesquisarProduto = QtWidgets.QPushButton(self.layoutWidget)
        self.btnPesquisarProduto.setMaximumSize(QtCore.QSize(100, 40))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.btnPesquisarProduto.setFont(font)
        icon1 = QtGui.QIcon()
        icon1.addPixmap(QtGui.QPixmap("lupa.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.btnPesquisarProduto.setIcon(icon1)
        self.btnPesquisarProduto.setIconSize(QtCore.QSize(16, 16))
        self.btnPesquisarProduto.setObjectName("btnPesquisarProduto")
        self.horizontalLayout.addWidget(self.btnPesquisarProduto)
        self.btnLimpar = QtWidgets.QPushButton(self.layoutWidget)
        self.btnLimpar.setMaximumSize(QtCore.QSize(100, 40))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.btnLimpar.setFont(font)
        icon2 = QtGui.QIcon()
        icon2.addPixmap(QtGui.QPixmap("borracha.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.btnLimpar.setIcon(icon2)
        self.btnLimpar.setIconSize(QtCore.QSize(16, 16))
        self.btnLimpar.setObjectName("btnLimpar")
        self.horizontalLayout.addWidget(self.btnLimpar)
        self.layoutWidget1 = QtWidgets.QWidget(self.centralwidget)
        self.layoutWidget1.setGeometry(QtCore.QRect(140, 380, 420, 42))
        self.layoutWidget1.setObjectName("layoutWidget1")
        self.horizontalLayout_2 = QtWidgets.QHBoxLayout(self.layoutWidget1)
        self.horizontalLayout_2.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout_2.setObjectName("horizontalLayout_2")
        self.btnExcluirProduto = QtWidgets.QPushButton(self.layoutWidget1)
        font = QtGui.QFont()
        font.setPointSize(10)
        self.btnExcluirProduto.setFont(font)
        icon3 = QtGui.QIcon()
        icon3.addPixmap(QtGui.QPixmap("deletar.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.btnExcluirProduto.setIcon(icon3)
        self.btnExcluirProduto.setIconSize(QtCore.QSize(30, 30))
        self.btnExcluirProduto.setObjectName("btnExcluirProduto")
        self.horizontalLayout_2.addWidget(self.btnExcluirProduto)
        self.gridLayout = QtWidgets.QGridLayout()
        self.gridLayout.setObjectName("gridLayout")
        self.btnAlterarProduto = QtWidgets.QPushButton(self.layoutWidget1)
        font = QtGui.QFont()
        font.setPointSize(10)
        self.btnAlterarProduto.setFont(font)
        icon4 = QtGui.QIcon()
        icon4.addPixmap(QtGui.QPixmap("alterar produto.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.btnAlterarProduto.setIcon(icon4)
        self.btnAlterarProduto.setIconSize(QtCore.QSize(30, 30))
        self.btnAlterarProduto.setObjectName("btnAlterarProduto")
        self.gridLayout.addWidget(self.btnAlterarProduto, 0, 0, 1, 1)
        self.btnCadastrarProduto = QtWidgets.QPushButton(self.layoutWidget1)
        font = QtGui.QFont()
        font.setPointSize(10)
        self.btnCadastrarProduto.setFont(font)
        icon5 = QtGui.QIcon()
        icon5.addPixmap(QtGui.QPixmap("cadastrar produtos 2.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.btnCadastrarProduto.setIcon(icon5)
        self.btnCadastrarProduto.setIconSize(QtCore.QSize(30, 30))
        self.btnCadastrarProduto.setObjectName("btnCadastrarProduto")
        self.gridLayout.addWidget(self.btnCadastrarProduto, 0, 1, 1, 1)
        self.horizontalLayout_2.addLayout(self.gridLayout)
        FormProdutos.setCentralWidget(self.centralwidget)
        self.statusbar = QtWidgets.QStatusBar(FormProdutos)
        self.statusbar.setObjectName("statusbar")
        FormProdutos.setStatusBar(self.statusbar)

        self.retranslateUi(FormProdutos)
        self.btnLimpar.clicked.connect(self.editPesquisaProduto.clear)
        self.btnCadastrarProduto.clicked.connect(self.click_btn_cadastrar_produto)
        self.btnPesquisarProduto.clicked.connect(self.click_btn_pesquisar_produto)
        self.btnAlterarProduto.clicked.connect(self.click_btn_alterar_produto)
        self.btnExcluirProduto.clicked.connect(self.click_btn_excluir_produto)


        QtCore.QMetaObject.connectSlotsByName(FormProdutos)
        FormProdutos.setTabOrder(self.editPesquisaProduto, self.btnPesquisarProduto)
        FormProdutos.setTabOrder(self.btnPesquisarProduto, self.btnLimpar)
        FormProdutos.setTabOrder(self.btnLimpar, self.tableProduto)
        FormProdutos.setTabOrder(self.tableProduto, self.btnAlterarProduto)
        FormProdutos.setTabOrder(self.btnAlterarProduto, self.btnCadastrarProduto)

    def retranslateUi(self, FormProdutos):
        _translate = QtCore.QCoreApplication.translate
        FormProdutos.setWindowTitle(_translate("FormProdutos", "Produtos"))
        self.label.setText(_translate("FormProdutos", "Produto:"))
        item = self.tableProduto.horizontalHeaderItem(0)
        item.setText(_translate("FormProdutos", "Cód. Produto"))
        item = self.tableProduto.horizontalHeaderItem(1)
        item.setText(_translate("FormProdutos", "Descrição"))
        item = self.tableProduto.horizontalHeaderItem(2)
        item.setText(_translate("FormProdutos", "Tipo"))
        item = self.tableProduto.horizontalHeaderItem(3)
        item.setText(_translate("FormProdutos", "Preço"))

        self.btnPesquisarProduto.setText(_translate("FormProdutos", "Pesquisar"))
        self.btnLimpar.setText(_translate("FormProdutos", "Limpar"))
        self.btnExcluirProduto.setText(_translate("FormProdutos", "Exclur Produto"))
        self.btnAlterarProduto.setText(_translate("FormProdutos", "Alterar Produto"))
        self.btnCadastrarProduto.setText(_translate("FormProdutos", "Cadastrar Produto"))

if __name__ == '__main__':
    import sys
    app = QApplication(sys.argv)
    MainWindow = QMainWindow()
    ui = Ui_FormProdutos()
    ui.setupUi(MainWindow)
    MainWindow.show()
    sys.exit(app.exec_())